framework/src/modules/chain/bft/bft.js

```diff
@@ -14,273 +14,136 @@
-			maxPreVoteHeight: 0,
-			maxPreCommitHeight: 0,
-		};
-
-		// Get first block of the round when delegate was active
-		const delegateMinHeightActive =
-			(header.activeSinceRound - 1) * this.activeDelegates + 1;
-
-		// If delegate is new then first block of the round will be considered
-		// if it forged before then we probably have the last commit height
-		// delegate can't pre-commit and block before above mentioned conditions
-		const minPreCommit = Math.max(
-			delegateMinHeightActive,
-			delegateState.maxPreCommitHeight + 1
+	async init() {
+		const finalizedHeight = await this._initConsensusManager();
+		const lastBlockHeight = await this._getLastBlockHeight();
+
+		const loadFromHeight = Math.max(
+			finalizedHeight,
+			lastBlockHeight - this.constants.activeDelegates * 2,
+			this.constants.startingHeight
 		);
 
-		// Delegate can't pre-commit the blocks on tip of the chain
-		const maxPreCommitHeight = header.height - 1;
-
-		for (let j = minPreCommit; j <= maxPreCommitHeight; j++) {
-			// Add pre-commit if threshold is reached
-			if (this.preVotes[j] >= this.preVoteThreshold) {
-				// Increase the pre-commit for particular height
-				this.preCommits[j] = (this.preCommits[j] || 0) + 1;
-
-				// Keep track of the last pre-commit point
-				delegateState.maxPreCommitHeight = j;
-			}
-		}
-
-		// Check between height of first block of the round when delegate was active
-		// Or one step ahead where it forged the last block
-		// Or one step ahead where it left the last pre-vote
-		// Or maximum 3 rounds backward
-		const minPreVoteHeight = Math.max(
-			delegateMinHeightActive,
-			header.maxHeightPreviouslyForged + 1,
-			delegateState.maxPreVoteHeight + 1,
-			header.height - this.processingThreshold
-		);
-		// Pre-vote upto current block height
-		const maxPreVoteHeight = header.height;
-		for (let j = minPreVoteHeight; j <= maxPreVoteHeight; j++) {
-			this.preVotes[j] = (this.preVotes[j] || 0) + 1;
-		}
-		// Update delegate state
-		delegateState.maxPreVoteHeight = maxPreVoteHeight;
-
-		// Set the delegate state
-		this.state[delegatePublicKey] = delegateState;
+		await this.loadBlocks({
+			fromHeight: loadFromHeight,
+			tillHeight: lastBlockHeight,
+		});
 	}
 
 	/**
-	 * Update the pre-voted confirmed and finalized height
+	 * Initialize the consensus manager and return the finalize height
+	 *
+	 * @return {Promise<number>} - Return the finalize height
+	 * @private
 	 */
-	updatePreVotedAndFinalizedHeight() {
-		if (this.headers.length === 0) {
-			return;
-		}
-
-		const higherPairVoted = Object.entries(this.preVotes)
-			.reverse()
-			.find(pair => pair[1] >= this.preVoteThreshold);
-
-		this.prevotedConfirmedHeight = higherPairVoted
-			? parseInt(higherPairVoted[0])
-			: this.prevotedConfirmedHeight;
-
-		const higherPairCommitted = Object.entries(this.preCommits)
-			.reverse()
-			.find(pair => pair[1] >= this.preCommitThreshold);
+	async _initConsensusManager() {
```

Comment:
initialization returning finalized height seems weird to me.
Probably getting finalized height and intializing consensus manager should be separated

Reply:
Yeah sure.